### PR TITLE
core: fix putMany for redundant DistanceRangeMaps

### LINKED ISSUE #6726: Core: DistanceRangeMap.putMany() doesn't merge equal values for consecutive intervals
### What happened?

See tests in https://github.com/osrd-project/osrd/compare/peb/core/fix_distance_range_put_many
The merge when extremities and values are the same is not done by `putMany()`, leading to a different behavior than expected, and than `put()`.

### What did you expect to happen?

Have `putMany()` merging consecutive ranges with identical values.

### How can we reproduce it (as minimally and precisely as possible)?

Use tests available in `peb/core/fix_distance_range_put_many` branch

### What operating system, browser and environment are you using?

- WSL
- Gradle
- Java 17

### OSRD version (top right corner `Account` button > `Informations`)

dfcbaad3d49b68c917642fb4eaad0f05605d10f3

### PR DESCRIPTION
**Pull request by shenriotpro**
fix https://github.com/osrd-project/osrd/issues/6726

### CODE REVIEW

File: core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/DistanceRangeMapImpl.kt [hunk 1]
```diff
@@ -46,42 +85,46 @@ data class DistanceRangeMapImpl<T>(
...
-            if (entryQueue.contains(index)) entryQueue.remove(index) else entryQueue.add(index)
 
+        // Compute bounds and values.
+        // E.g. nonEmptyBoundValues = [(0, 1), (5, 1), (10, null)]
+        val nonEmptyBoundValues = mutableListOf<Pair<Distance, T?>>()
+        for ((bound, indices) in entriesByBound) {
+            for (index in indices) {
+                // Update relevant entries. PriorityQueue only guarantees linear time for contains
+                // and
+                // remove, an optimized heap could be helpful.
```
Comment (bougue-pe) [inline core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/DistanceRangeMapImpl.kt]: nit on ktfmt:
```suggestion
                // Update relevant entries. PriorityQueue only guarantees linear time for contains
                // and remove, an optimized heap could be helpful.
```

File: core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/DistanceRangeMapImpl.kt [hunk 2]
```diff
@@ -33,7 +33,46 @@ data class DistanceRangeMapImpl<T>(
...
+            }
+            result.add(Pair(nextKey, nextValues))
+            return result
+        }
+
+        fun <U, V> groupBySecond(pairs: List<Pair<U, V>>): List<Pair<V, List<U>>> {
+            val result = mutableListOf<Pair<V, List<U>>>()
+            if (pairs.isEmpty()) return result
+            var nextKey = pairs[0].second
+            var nextValues = mutableListOf(pairs[0].first)
```
Comment (bougue-pe) [inline core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/DistanceRangeMapImpl.kt]: minor: Can't you mutualize the code by reversing the pairs first or providing a specific getKey/getValue ?
Not sure about the tradeoff on perf and readability.

**Review by bougue-pe — APPROVED** [review]
LGTM, the comments help understand at least the (edgy) case 👍
And the perf check is still OK.

File: core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/DistanceRangeMapImpl.kt [hunk 3]
```diff
@@ -33,7 +33,36 @@ data class DistanceRangeMapImpl<T>(
      * structure (RangeMaps lib).
      */
     override fun putMany(entries: List<DistanceRangeMap.RangeMapEntry<T>>) {
+        // Unfortunately, built-in groupBy doesn't consider the original order, here we want
+        // consecutive values.
+        fun <U, V> groupByFirst(pairs: List<Pair<U, V>>): List<Pair<U, List<V>>> {
```
Comment (multun) [inline core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/DistanceRangeMapImpl.kt]: Can you make it clearer that this variant of groupBy only groups consecutive values? groupBy, by convention, always refers to the unordered variant

Comment (multun) [inline core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/DistanceRangeMapImpl.kt]:
```suggestion
        fun <U, V> groupByConsecutiveFirst(pairs: List<Pair<U, V>>): List<Pair<U, List<V>>> {
```

**Review by multun — APPROVED** [review]

**Pull request merged**